File: Python/Algorithms/1370.py
from collections import defaultdict


class Solution:
    def sortString(self, s):
        # Keep sorted list of elements,
        sl = sorted(set(s))
        len_ = len
        # edge case, return early.
        if len_(sl) == 1:
            return s

        # keep reversed list of sorted elements.
        revsl = [*reversed(sl)]
        # hold counts for each character.
        counts = defaultdict(int)
        for i in s:
            counts[i] += 1

        # hold chars
        res = []
        append = res.append
        get, pop = counts.get, counts.pop

        # Flags for handling in which direction we loop and when to stop
        toggle, pending = False, True
        while pending:
            for i in sl if not toggle else revsl:
                if i in counts:
                    # if equal to one, final time we see it.
                    # remove it and continue.
                    if get(i) == 1:
                        pop(i)
                        # ok, stop, we can add manually.
                        if len_(counts) == 1:
                            pending = False
                    # it's count is > 1. Decrease it and continue.
                    else:
                        counts[i] -= 1
                    append(i)
            # Toggle direction in which we iterate through characters.
            toggle = not toggle

        # add any left-over character if it exists.
        if len_(counts):
            c, count = counts.popitem()
            res.extend([c for _ in range(count)])

        return "".join(res)

File: Javascript/Algorithms/1370.js
/**
 * @param {string} s
 * @return {string}
 */
let sortString = s => {
    const sorted = Array.from(new Set(s));
    sorted.sort()
    
    // Early bails
    if (sorted.length == 1){
        return s;
    }
    if (sorted.length == s.length){
        return sorted.join("");
    }
    
    // Create reversed array.
    const reversed = Array.from(sorted);
    reversed.reverse();
    // Build counts.
    const counts = new Map;
    for (let i of s){
        if (counts.has(i)){
            counts.set(i, counts.get(i) + 1);
        } else {
            counts.set(i, 1);
        }
    }
    
    let result = [];
    let [toggle, pending] = [false, true];
    // While counts.size > 1, keep going.
    while (pending){
        let it = toggle ? reversed: sorted;
        for (let i of it){
            if (counts.has(i)){
                let val = counts.get(i);
                // Delete counts[i] if there's only one of it left
                if (val == 1){
                    counts.delete(i)
                    // break 
                    if (counts.size == 1){
                        pending = !pending;
                    }
                // Decrease count
                } else {
                    counts.set(i, val - 1);
                }
                result.push(i);
            }
        }
        toggle = !toggle;
    }
    // Add remaining offset if it exists.
    let offset = "";
    if (counts.size){
        let [char, reps] = counts.entries().next().value;
        offset = char.repeat(reps);
    }
    result = result.join("")
    result += offset;
    return result;
};
